File: py/file_reader/stack.py
from typing import TypeVar, Generic

T = TypeVar('T')


class Stack(Generic[T]):
    """
    Generic stack data structure implementation
    """

    def __init__(self) -> None:
        self.__items: list[T] = []

    def append(self, item) -> None:
        self.__items.append(item)

    def pop(self) -> T:
        return self.__items.pop()

    def is_empty(self) -> bool:
        return not self.__items

    def peek(self) -> T:
        return self.__items[-1]

File: ts/utils/Stack.ts
// LIFO data structure, stack is commonly used to avoid resursion
export class Stack<T> {
  #items: T[] = [];

  isEmpty(): Boolean {
    return !this.#items.length;
  };

  push(item: T): void {
    this.#items.push(item);
  };

  peek(): T {
    if (this.isEmpty()) {
      throw new Error('accessing empty stack');
    }
    return this.#items.at(-1)!;
  };

  pop(): T {
    if (this.isEmpty()) {
      throw new Error('popping from empty stack');
    }
    return this.#items.pop()!;
  };
};
